// botasaurus-desktop-api/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiException = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const utils_1 = require("./utils");
class ApiException extends Error {
    constructor(message) {
        super(message);
        this.name = "ApiException";
    }
}
exports.ApiException = ApiException;
function _createFilename(path) {
    return "output/responses/" + path + ".json";
}
function raiseIfBadException(response) {
    if (400 <= response.status && response.status < 500) {
        const data = response.data;
        const message = data === null || data === void 0 ? void 0 : data.message;
        if (message) {
            throw new ApiException(message);
        }
    }
}
// Helper to normalize path similar to Python's os.path.normpath
function normalizePath(path) {
    return path
        .split("/")
        .filter(Boolean)
        .join("/")
        .replace(/\/+/g, "/");
}
function cleanBasePath(apiBasePath) {
    if (typeof apiBasePath === "string") {
        let path = apiBasePath ? normalizePath(apiBasePath) : "";
        if (path === ".") {
            path = "";
        }
        else if (path) {
            if (!path.startsWith("/")) {
                path = "/" + path;
            }
            if (path.endsWith("/")) {
                path = path.slice(0, -1);
            }
        }
        return path;
    }
    return '';
}
class Api {
    constructor({ apiUrl, createResponseFiles = false, apiBasePath = "", enableCache } = { createResponseFiles: true }) {
        Object.defineProperty(this, "_apiUrl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_apiBasePath", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_createResponseFiles", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_enableCache", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "DEFAULT_API_URL", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "http://localhost:8000"
        });
        /**
         * Initializes the API client with a specified server URL, base path, caching options, and an option to create response files.
         *
         * @param apiUrl The base URL for the API server. If not specified, defaults to "http://localhost:8000".
         * @param createResponseFiles Indicates if the client should create response files for each API call. This is useful for debugging or development purposes. Defaults to true.
         * @param apiBasePath Base path to prefix API endpoints (e.g., "/v1").
         * @param enableCache Enable or disable caching for all API requests. Overrides the server's cache setting.
         */
        this._apiUrl = apiUrl
            ? (0, utils_1.removeAfterFirstSlash)(apiUrl)
            : this.DEFAULT_API_URL;
        this._createResponseFiles = createResponseFiles;
        this._apiBasePath = cleanBasePath(apiBasePath);
        this._enableCache = enableCache;
        // Check if API is running (note: this is synchronous in the original, but should be async in TS)
        this.isApiRunning().then((isRunning) => {
            if (!isRunning) {
                throw new ApiException(`API at ${this._apiUrl}${this._apiBasePath} is not running. Please check if the API is up and running.`);
            }
        });
    }
    _writeJson(filename, data) {
        /**
         * Writes data to a JSON file specified by the filename. This method only runs if the createResponseFiles flag is True.
         *
         * @param filename The filename for the JSON file to be created.
         * @param data The data to be written to the file.
         */
        if (this._createResponseFiles) {
            const path = _createFilename(filename);
            (0, utils_1.writeJsonResponse)(path, data);
            console.log(`View ${filename} response at: ./${path}`);
        }
    }
    _makeApiUrl(path) {
        return `${this._apiUrl}${this._apiBasePath}${path === '' && this._apiBasePath ? "" : "/"}${path}`;
    }
    _getCacheParams() {
        return typeof this._enableCache === 'boolean' ? { enable_cache: this._enableCache } : {};
    }
    async isApiRunning() {
        /**
         * Checks if the API is running by performing a health check on the "/api" endpoint.
         *
         * @return True if the health check is successful, otherwise False.
         */
        try {
            const response = await axios_1.default.get(this._makeApiUrl(""));
            return response.status === 200;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error) && error.code === "ECONNREFUSED") {
                throw new ApiException(`API at ${this._apiUrl}${this._apiBasePath} is not running. 
Check the network connection, or verify if the API is running on a different endpoint. In case the API is running on a different endpoint, you can pass the endpoint as follows:
const api = new Api({ apiUrl: 'https://example.com' })`);
            }
            throw error;
        }
    }
    async createAsyncTask({ data, scraperName }) {
        /**
         * Submits an asynchronous task to the server.
         *
         * @param data The data to be received by the scraper.
         * @param scraperName The name of the scraper to use for the task. If not provided, the server will use the default scraper.
         * @return The created task object.
         */
        const url = this._makeApiUrl("tasks/create-task-async");
        const payload = {
            data,
            scraper_name: scraperName,
            ...this._getCacheParams(),
        };
        try {
            const response = await axios_1.default.post(url, payload);
            this._writeJson("create_async_task", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async createSyncTask({ data, scraperName }) {
        /**
         * Submits a synchronous task to the server.
         *
         * @param data The data to be received by the scraper.
         * @param scraperName The name of the scraper to use for the task. If not provided, the server will use the default scraper.
         * @return The created task object.
         */
        const url = this._makeApiUrl("tasks/create-task-sync");
        const payload = {
            data,
            scraper_name: scraperName,
            ...this._getCacheParams(),
        };
        try {
            const response = await axios_1.default.post(url, payload);
            this._writeJson("create_sync_task", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async createAsyncTasks({ dataItems, scraperName }) {
        /**
         * Submits multiple asynchronous tasks to the server in a batch.
         *
         * @param dataItems A list of data items to be processed by the scraper.
         * @param scraperName The name of the scraper to use for the tasks. If not provided, the server will use the default scraper.
         * @return A list of created task objects.
         */
        const url = this._makeApiUrl("tasks/create-task-async");
        const payload = dataItems.map((data) => ({
            data,
            scraper_name: scraperName,
            ...this._getCacheParams()
        }));
        try {
            const response = await axios_1.default.post(url, payload);
            this._writeJson("create_async_tasks", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async createSyncTasks({ dataItems, scraperName }) {
        /**
         * Submits multiple synchronous tasks to the server in a batch and waits for the results.
         *
         * @param dataItems A list of data items to be processed by the scraper.
         * @param scraperName The name of the scraper to use for the tasks. If not provided, the server will use the default scraper.
         * @return A list of created task objects, each with its processing result.
         */
        const url = this._makeApiUrl("tasks/create-task-sync");
        const payload = dataItems.map((data) => ({
            data,
            scraper_name: scraperName,
            ...this._getCacheParams()
        }));
        try {
            const response = await axios_1.default.post(url, payload);
            this._writeJson("create_sync_tasks", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async getTasks({ page = 1, perPage, withResults = false, parentTaskId, status } = { page: 1, withResults: true }) {
        /**
         * Fetches tasks from the server, with optional result inclusion, pagination, and filtering.
         *
         * @param page The page number for pagination.
         * @param perPage The number of tasks to return per page.
         * @param withResults Whether to include the task results in the response.
         * @param parentTaskId Filter tasks by parent task ID.
         * @param status Filter tasks by status (pending, in_progress, completed, failed, aborted).
         * @return A dictionary containing the task results and pagination information.
         */
        const url = this._makeApiUrl("tasks");
        const params = {
            with_results: withResults,
        };
        if (page !== undefined && page !== null)
            params.page = page;
        if (perPage !== undefined && perPage !== null)
            params.per_page = perPage;
        if (parentTaskId !== undefined && parentTaskId !== null)
            params.parent_task_id = parentTaskId;
        if (status !== undefined && status !== null)
            params.status = status;
        try {
            const response = await axios_1.default.get(url, { params });
            if (this._createResponseFiles) {
                const hasManyPages = response.data.total_pages > 1;
                const filename = hasManyPages
                    ? `get_tasks-page-${page}`
                    : "get_tasks";
                const msg = hasManyPages
                    ? `get_tasks, page ${page}`
                    : "get_tasks";
                const path = _createFilename(filename);
                (0, utils_1.writeJsonResponse)(path, response.data);
                console.log(`View ${msg} response at: ./${path}`);
            }
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async getFailedTasks({ page = 1, perPage, withResults = false, parentTaskId } = { page: 1, withResults: true }) {
        return this.getTasks({ page, perPage, withResults, parentTaskId, status: 'failed' });
    }
    async getTask(taskId) {
        /**
         * Retrieves a specific task by ID.
         *
         * @param taskId The ID of the task to retrieve.
         * @return The task object.
         */
        const url = this._makeApiUrl(`tasks/${taskId}`);
        try {
            const response = await axios_1.default.get(url);
            this._writeJson("get_task", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async getTaskResults({ taskId, filters, sort, view, page = 1, perPage }) {
        /**
         * Retrieves the results of a specific task.
         *
         * @param taskId The ID of the task.
         * @param filters A dictionary of filters to apply to the task results, optional.
         * @param sort The sort to apply to the task results, optional.
         * @param view The view to apply to the task results, optional.
         * @param page The page number to retrieve, default is 1.
         * @param perPage The number of results to return per page. If perPage is not provided, all results are returned, optional.
         * @return A dictionary containing the task results and pagination information if page and perPage are provided.
         */
        const url = this._makeApiUrl(`tasks/${taskId}/results`);
        const payload = { page };
        if (filters)
            payload.filters = filters;
        if (sort)
            payload.sort = sort;
        if (view)
            payload.view = view;
        if (perPage)
            payload.per_page = perPage;
        try {
            const response = await axios_1.default.post(url, payload);
            if (this._createResponseFiles) {
                const hasManyPages = response.data.total_pages > 1;
                const filename = hasManyPages
                    ? `get_task_results-page-${page}`
                    : "get_task_results";
                const msg = hasManyPages
                    ? `get_task_results, page ${page}`
                    : "get_task_results";
                const path = _createFilename(filename);
                (0, utils_1.writeJsonResponse)(path, response.data);
                console.log(`View ${msg} response at: ./${path}`);
            }
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async get(scraperName, data) {
        /**
         * Makes a direct GET request to the scraper endpoint, bypassing the task creating and scheduling system.
         *
         * @param scraperName The name of the scraper, which corresponds to the route path (e.g., "scrape-heading-task").
         * @param data The data to be received by the scraper.
         * @return The result from the scraper.
         */
        if (scraperName.startsWith('/')) {
            scraperName = scraperName.slice(1);
        }
        const url = this._makeApiUrl(scraperName);
        const params = {
            ...data,
            ...this._getCacheParams()
        };
        try {
            const response = await axios_1.default.get(url, { params });
            const fileScraperName = scraperName.replaceAll('/', '-');
            this._writeJson(`${fileScraperName}-result`, response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async downloadTaskResults({ taskId, format, filters, sort, view, convertToEnglish = true }) {
        /**
         * Downloads the results of a specific task in the specified format.
         *
         * @param taskId The ID of the task.
         * @param format The format to download the task results in. Available formats are "json", "csv", "excel". The default format is "json".
         * @param filters A dictionary of filters to apply to the task results, optional.
         * @param sort The sort to apply to the task results, optional.
         * @param view The view to apply to the task results, optional.
         * @param convertToEnglish Whether to convert the task results to English, default is True, optional.
         * @return A tuple containing the downloaded content as a buffer and the filename.
         */
        const url = this._makeApiUrl(`tasks/${taskId}/download`);
        const payload = { convert_to_english: convertToEnglish };
        if (format)
            payload.format = format;
        if (filters)
            payload.filters = filters;
        if (sort)
            payload.sort = sort;
        if (view)
            payload.view = view;
        try {
            const response = await axios_1.default.post(url, payload, {
                responseType: "arraybuffer",
            });
            let buffer;
            try {
                buffer = Buffer.from(response.data, 'utf8');
            }
            catch (bufferError) {
                throw new ApiException("Failed to create buffer from response data. Ensure the response data is valid.");
            }
            // @ts-ignore
            const filename = (0, utils_1.getFilenameFromResponseHeaders)(response);
            if (this._createResponseFiles) {
                const path = (0, utils_1.writeFileResponse)("output/responses/", filename, buffer);
                console.log(`View downloaded file at: ./${path}`);
            }
            return { buffer, filename };
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async abortTask(taskId) {
        /**
         * Aborts a specific task.
         *
         * @param taskId The ID of the task to abort.
         * @return A success message.
         */
        const url = this._makeApiUrl(`tasks/${taskId}/abort`);
        try {
            const response = await axios_1.default.patch(url, {}, {
                headers: {
                    'Content-Type': 'application/json'
                }
            });
            this._writeJson("abort_task", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async deleteTask(taskId) {
        /**
         * Deletes a specific task.
         *
         * @param taskId The ID of the task to delete.
         * @return A success message.
         */
        const url = this._makeApiUrl(`tasks/${taskId}`);
        try {
            const response = await axios_1.default.delete(url);
            this._writeJson("delete_task", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async deleteTasks({ taskIds }) {
        /**
         * Bulk deletes tasks.
         *
         * @param taskIds A list of task IDs to be deleted.
         * @return A success message.
         */
        const url = this._makeApiUrl("tasks/bulk-delete");
        const payload = { task_ids: taskIds };
        try {
            const response = await axios_1.default.post(url, payload);
            this._writeJson("delete_tasks", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async abortTasks({ taskIds }) {
        /**
         * Bulk aborts tasks.
         *
         * @param taskIds A list of task IDs to be aborted.
         * @return A success message.
         */
        const url = this._makeApiUrl("tasks/bulk-abort");
        const payload = { task_ids: taskIds };
        try {
            const response = await axios_1.default.post(url, payload);
            this._writeJson("abort_tasks", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
    async retryTasks({ taskIds }) {
        /**
         * Bulk retries failed tasks.
         *
         * @param taskIds A list of task IDs to be retried.
         * @return A success message.
         */
        const url = this._makeApiUrl("tasks/bulk-retry");
        const payload = { task_ids: taskIds };
        try {
            const response = await axios_1.default.post(url, payload);
            this._writeJson("retry_tasks", response.data);
            return response.data;
        }
        catch (error) {
            if (axios_1.default.isAxiosError(error)) {
                raiseIfBadException(error.response);
            }
            throw error;
        }
    }
}
exports.default = Api;
