// botasaurus-desktop-api/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeAfterFirstSlash = removeAfterFirstSlash;
exports.relativePath = relativePath;
exports.createDirectoryIfNotExists = createDirectoryIfNotExists;
exports.createOutputDirectoryIfNotExists = createOutputDirectoryIfNotExists;
exports.writeJsonResponse = writeJsonResponse;
exports.getFilenameFromResponseHeaders = getFilenameFromResponseHeaders;
exports.writeFileResponse = writeFileResponse;
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const path_1 = tslib_1.__importDefault(require("path"));
/**
 * Removes everything after (and including) the first single slash (/) in a string,
 * but skips over '//' (double slashes), which are used in protocols like 'http://'.
 * This is primarily used to extract the base URL (protocol + host) from a full URL.
 *
 * Example:
 *   removeAfterFirstSlash("https://example.com/api/v1") // "https://example.com"
 *   removeAfterFirstSlash("http://foo/bar/baz") // "http://foo"
 *   removeAfterFirstSlash("localhost:8000/api/v1") // "localhost:8000"
 *   removeAfterFirstSlash("https://example.com") // "https://example.com"
 */
function removeAfterFirstSlash(inputString) {
    let i = 0;
    const strLen = inputString.length;
    while (i < strLen) {
        const char = inputString[i];
        if (char === '/') {
            // Check for double slash (e.g., 'http://')
            if (i + 1 < strLen && inputString[i + 1] === '/') {
                i += 2;
                continue;
            }
            else {
                // Single slash: trim here
                return inputString.substring(0, i);
            }
        }
        i++;
    }
    // No single slash found (or only part of protocol), return the whole input
    return inputString;
}
function relativePath(targetPath, goback = 0) {
    const levels = Array(Math.max(0, goback - 1)).fill('..');
    return path_1.default.resolve(process.cwd(), ...levels, targetPath.trim());
}
function createDirectoryIfNotExists(passedPath) {
    const dirPath = relativePath(passedPath, 0);
    if (!fs_1.default.existsSync(dirPath)) {
        fs_1.default.mkdirSync(dirPath, { recursive: true });
    }
}
let outputDirectoryCreated = false;
function createOutputDirectoryIfNotExists() {
    // Check if output directory exists, if not, create it
    if (!outputDirectoryCreated) {
        createDirectoryIfNotExists("output");
        createDirectoryIfNotExists("output/responses");
        outputDirectoryCreated = true;
    }
}
function writeJsonResponse(filePath, data, indent = 4) {
    createOutputDirectoryIfNotExists();
    fs_1.default.writeFileSync(filePath, JSON.stringify(data, null, indent), { encoding: 'utf-8' });
}
function getFilenameFromResponseHeaders(response) {
    const contentDisposition = response.headers.get('Content-Disposition');
    if (!contentDisposition)
        throw new Error('Content-Disposition header not found');
    const filenamePart = contentDisposition.split('filename=')[1];
    if (!filenamePart)
        throw new Error('Filename not found in Content-Disposition');
    return filenamePart.trim().replace(/^"|"$/g, '');
}
function writeFileResponse(filePath, filename, content) {
    createOutputDirectoryIfNotExists();
    const fullPath = path_1.default.join(filePath, filename);
    fs_1.default.writeFileSync(fullPath, content);
    return fullPath;
}
